### PR TITLE
DM-1876 OpenTelemetry PlatformMetricExporter

### PR DESCRIPTION
**Pull request by ashhSAG93**
The PR describes how an OpenTelemetry Exporter can be implemented for Cumulocity Platform, And some of the limitations associated with gauge metrics and closing of callbacks, due to which we have decided to keep this implementation in draft state for now and explore Micrometer or other options for metrics instrumentation. The PR is in draft state and not to be reviewed

### CODE REVIEW

File: opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/common/OpenTelemetryExporterStrategy.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.cumulocity.exporters.common;
+
+import io.opentelemetry.api.OpenTelemetry;
+
+public interface OpenTelemetryExporterStrategy {
```
Comment (ashhSAG93) [inline opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/common/OpenTelemetryExporterStrategy.java:5]: This is the initial implementation of a Strategy pattern enabling users to select from multiple exporters

File: opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        for(int gaugeIndex = 0; gaugeIndex < gaugeMetrics.size(); gaugeIndex++) {
+            gaugePointSize += gaugeMetrics.get(gaugeIndex).getDoubleGaugeData().getPoints().size();
+        }
+        for(MetricData metric : otherMetrics) {
+            exportNonGaugeMetrics(metric);
+        }
+        for(MetricData metric: gaugeMetrics) {
+            exportGaugeMetrics(metric);
+        }
+        applicationEventPublisher.publishEvent(new ExportCompletedEvent("exportCompletedEvent", gaugePointSize));
```
Comment (ashhSAG93) [inline opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java:87]: Limitation with Gauge Metric Instrumentation: The callbacks registered on instrumented the metric arent closed on export. Hence we need to publish an event, which when received by the controller/service, instrumenting the metrics, closes the list of gauges stored on instrumentation

File: opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+@Slf4j
+public class PlatformMetricExporter implements MetricExporter {
+
+
+    private RestConnector restConnector;
+
+    private ApplicationEventPublisher applicationEventPublisher;
+
+    public PlatformMetricExporter(RestConnector restConnector, ApplicationEventPublisher applicationEventPublisher) {
```
Comment (ashhSAG93) [inline opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java:31]: This class is not made a component as the TenantScope required for the RestConnector is lost when the Exporter is invoked by the running PeriodicMetricReader Scheduler. Hence this constructor is created, which can be used in the PlatformOpenTelemetryConfiguration

File: opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        log.info("Sending metric for tenant {} named {} whose value is {} and whose type is {}", restConnector.getPlatformParameters().getTenantId(), name, value, metric.getType().name());
+        MicroserviceBillingRepresentation representation = new MicroserviceBillingRepresentation();
+        representation.setDateTime(DateTime.now());
+        representation.set(value, name);
+        restConnector.post("/application/currentApplication/billing",
+                APPLICATION_JSON_TYPE, representation);
+    }
+
+    private void exportGaugeMetrics(MetricData metric) {
+        Collection<DoublePointData> points = metric.getDoubleGaugeData().getPoints();
```
Comment (ashhSAG93) [inline opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformMetricExporter.java:55]: Limitation of GaugeMetric:

Instrument name - Loriot Gauge Metrics

send 3 metrics in a window: 100 12:30 (attribute 12:30-12:30), 200 12:31(attribute 12:31-12:31), 300 12:32(attribute 12:32-12:32)

These are 3 points in the instrument loriot Gauge metrics

3 points = 3 callbacks

point data
100   12:30
200   12:31
300   12:32

The attributes are the way in which we can record the time at which the metric was actually instrumented. This logic creates a PointData-Time Map, using which we can get the most recent Metric sent and export that. If we do not send attributes in the metric points, only the first point is recorded, as the others are discarded due to the previous point already being recorded

File: opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformOpenTelemetryConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    private ApplicationEventPublisher applicationEventPublisher;
+
+    @Bean("platformOtel")
+    @TenantScope
+    public OpenTelemetry initOpenTelemetry() {
+//        PlatformExporter platformExporter = new PlatformExporter(contextService.getContext().getTenant(), applicationContext);
+        // Create an instance of PeriodicMetricReader and configure it
+        // to export via the logging exporter
+        Platform platform = PlatformBuilder.platform().withBaseUrl(baseURL).withForceInitialHost(true).withCredentials(contextService.getContext().toCumulocityCredentials()).build();
+        RestConnector restConnector = new RestConnector((PlatformParameters) platform, new ResponseParser());
```
Comment (ashhSAG93) [inline opentelemetry-metrics-exporters/src/main/java/com/cumulocity/exporters/platform/PlatformOpenTelemetryConfiguration.java:47]: To overcome TenantScope loss mentioned above.

**Comment by kaczynskid-sag** [comment]
Closing as outdated

**Pull request closed without merging**